pkg/controller/jaeger/deployment_test.go

```diff
@@ -0,0 +1,197 @@
+		depUpdated.Annotations = map[string]string{"key": "new-value"}
+
+		s := strategy.New().WithDeployments([]appsv1.Deployment{depUpdated})
+		return s
+	}
+
+	// test
+	_, err := r.Reconcile(reconcile.Request{NamespacedName: nsn})
+	assert.NoError(t, err)
+
+	// verify
+	persisted := &appsv1.Deployment{}
+	persistedName := types.NamespacedName{
+		Name:      depOriginal.Name,
+		Namespace: depOriginal.Namespace,
+	}
+	err = cl.Get(context.Background(), persistedName, persisted)
+	assert.Equal(t, "new-value", persisted.Annotations["key"])
+	assert.NoError(t, err)
+}
+
+func TestDeploymentDelete(t *testing.T) {
+	// prepare
+	nsn := types.NamespacedName{
+		Name: "TestDeploymentDelete",
+	}
+
+	depOriginal := appsv1.Deployment{}
+	depOriginal.Name = nsn.Name
+
+	objs := []runtime.Object{
+		v1alpha1.NewJaeger(nsn.Name),
+		&depOriginal,
+	}
+
+	r, cl := getReconciler(objs)
+	r.strategyChooser = func(jaeger *v1alpha1.Jaeger) strategy.S {
+		return strategy.S{}
+	}
+
+	// test
+	_, err := r.Reconcile(reconcile.Request{NamespacedName: nsn})
+	assert.NoError(t, err)
+
+	// verify
+	persisted := &appsv1.Deployment{}
+	persistedName := types.NamespacedName{
+		Name:      depOriginal.Name,
+		Namespace: depOriginal.Namespace,
+	}
+	err = cl.Get(context.Background(), persistedName, persisted)
+	assert.Empty(t, persisted.Name)
+	assert.Error(t, err) // not found
+}
+
+func TestDeploymentDeleteAfterCreate(t *testing.T) {
+	// prepare
+	nsn := types.NamespacedName{
+		Name: "TestDeploymentDeleteAfterCreate",
+	}
+
+	// the deployment to be deleted
+	depToDelete := appsv1.Deployment{}
+	depToDelete.Name = nsn.Name + "-delete"
+	depToDelete.Annotations = map[string]string{
+		"app.kubernetes.io/instance":   nsn.Name,
+		"app.kubernetes.io/managed-by": "jaeger-operator",
+	}
+	objs := []runtime.Object{v1alpha1.NewJaeger(nsn.Name), &depToDelete}
+
+	// the deployment to be created
+	dep := appsv1.Deployment{}
+	dep.Name = nsn.Name
+	dep.Status.Replicas = 2
+	dep.Status.ReadyReplicas = 1
+
+	r, cl := getReconciler(objs)
+	r.strategyChooser = func(jaeger *v1alpha1.Jaeger) strategy.S {
+		s := strategy.New().WithDeployments([]appsv1.Deployment{dep})
+		return s
+	}
+
+	// sanity check that the deployment to be removed is indeed there in the first place...
+	persistedDelete := &appsv1.Deployment{}
+	assert.NoError(t, cl.Get(context.Background(), types.NamespacedName{Name: depToDelete.Name, Namespace: depToDelete.Namespace}, persistedDelete))
+	assert.Equal(t, depToDelete.Name, persistedDelete.Name)
+
+	// test
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		// will block until it finishes, which should happen after the deployments
+		// have achieved stability and everything has been created/updated/deleted
+		_, err := r.Reconcile(reconcile.Request{NamespacedName: nsn})
+		assert.NoError(t, err)
+		wg.Done()
+	}()
+
+	// we assume that this sleep time is enough for the reconcile to reach the "wait" logic
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Could be this we avoided completelly?

Reply:
Unfortunately, no. I didn't have it before and it caused the tests to intermittently fail, as it's not certain that the "wait" part inside the reconcile loop would have started already.

Basically, what we want is the reconcile loop to run and create the deployment/cron job and wait until it reaches the stability / completes the job and we want the "waitFor" to execute at least once. *Then* we are ready to update the "available"/"succeeded" fields and the wait should finish, reaching the `wg.Done()` part.